pkg/snowflake/tag_association.go

```diff
@@ -0,0 +1,121 @@
+	ObjectSchema   sql.NullString `db:"OBJECT_SCHEMA"`
+	TagDatabase    sql.NullString `db:"TAG_DATABASE"`
+	TagId          sql.NullString `db:"TAG_ID"`
+	TagName        sql.NullString `db:"TAG_NAME"`
+	TagSchema      sql.NullString `db:"TAG_SCHEMA"`
+	TagValue       sql.NullString `db:"TAG_VALUE"`
+}
+
+// WithObjectId adds the name of the schema to the TagAssociationBuilder
+func (tb *TagAssociationBuilder) WithObjectName(objectName string) *TagAssociationBuilder {
+	tb.objectName = objectName
+	return tb
+}
+
+// WithObjectType adds the object type of the resource to add tag attachement to the TagAssociationBuilder
+func (tb *TagAssociationBuilder) WithObjectType(objectType string) *TagAssociationBuilder {
+	tb.objectType = objectType
+	return tb
+}
+
+// WithTagValue adds the name of the tag value to the TagAssociationBuilder
+func (tb *TagAssociationBuilder) WithTagValue(tagValue string) *TagAssociationBuilder {
+	tb.tagValue = tagValue
+	return tb
+}
+
+// GetTagDatabase returns the value of the tag database of TagAssociationBuilder
+func (tb *TagAssociationBuilder) GetTagDatabase() string {
+	return tb.databaseName
+}
+
+// GetTagName returns the value of the tag name of TagAssociationBuilder
+func (tb *TagAssociationBuilder) GetTagName() string {
+	return tb.schemaName
+}
+
+// GetTagSchema returns the value of the tag schema of TagAssociationBuilder
+func (tb *TagAssociationBuilder) GetTagSchema() string {
+	return tb.schemaName
+}
+
+// TagAssociation returns a pointer to a Builder that abstracts the DDL operations for a tag sssociation.
+//
+// Supported DDL operations are:
+//   - CREATE TAG
+//   - DROP TAG
+//
+// [Snowflake Reference](https://docs.snowflake.com/en/user-guide/object-tagging.html)
+func TagAssociation(tagID string) *TagAssociationBuilder {
+	parsedString := strings.Replace(tagID, "\"", "", -1)
+
+	var s []string
+	if strings.Contains(parsedString, "|") {
+		s = strings.Split(parsedString, "|")
+	} else if strings.Contains(parsedString, ".") {
+		s = strings.Split(parsedString, ".")
+	}
+	return &TagAssociationBuilder{
+		databaseName: s[0],
+		schemaName:   s[1],
+		tagName:      s[2],
+	}
+}
+
+// Create returns the SQL query that will set the tag on an object.
+func (tb *TagAssociationBuilder) Create() string {
+	return fmt.Sprintf(`ALTER %v "%v" SET TAG "%v"."%v"."%v" = '%v'`, tb.objectType, tb.objectName, tb.databaseName, tb.schemaName, tb.tagName, tb.tagValue)
+}
+
+// Drop returns the SQL query that will remove a tag from an object.
+func (tb *TagAssociationBuilder) Drop() string {
+	return fmt.Sprintf(`ALTER %v "%v" UNSET TAG "%v"."%v"."%v"`, tb.objectType, tb.objectName, tb.databaseName, tb.schemaName, tb.tagName)
+}
+
+func ListTagAssociations(tb *TagAssociationBuilder, db *sql.DB) ([]tagAssociation, error) {
+	stmt := fmt.Sprintf(`SELECT * FROM SNOWFLAKE.ACCOUNT_USAGE.TAG_REFERENCES WHERE TAG_NAME = '%v' AND DOMAIN = '%v' AND TAG_VALUE = '%v'`, tb.tagName, tb.objectType, tb.tagValue)
```

Comment:
@sfc-gh-swinkler Would you be so kind to provide a reasoning why the tags are read this way?

I see multiple potential issues here:
* minor nitpick: tags on deleted objects are not filtered (`OBJECT_DELETED IS NULL`)
* the query filters only on tag name instead of the fully qualified identifier (missing `TAG_DATABASE` and `TAG_SCHEMA`) - if I have two tags with the same name in different schemas, this would also select the other one
* why via the ACCOUNT_USAGE view? It would be possible to query the tags without any latency using [`SYSTEM$GET_TAG`](https://docs.snowflake.com/en/sql-reference/functions/system_get_tag.html) and compare the value afterwards with the state, if this is needed.
* Also, currently the query doesn't include the object name, so we are just querying if there is a tag on any object of the same domain with the given tag value set?

Reply:
@christophkreutzer thank you for pointing these out. i must say that i am not an expert at the Snowflake product, so I did not know about this SYSTEM$GET_TAG function. These would be good changes to make. If you can make a request with your account manager, then i can prioritize getting this done.

Reply:
@sfc-gh-swinkler thanks for the fast response! I created a first draft in PR #1261, and I'll let our account team know of the issues if that helps you taking a look at it.